src/main/kotlin/grpc/AuthenticationInterceptor.kt

```diff
@@ -0,0 +1,298 @@
+            logger.error { "Cookies map is missing from the gRPC context." }
+            throw SnowballRException.MissingContextException.MissingCookiesMap()
+        }
+
+        return cookies
+    }
+
+    /**
+     * Creates a `Set-Cookie` header string.
+     *
+     * @param name The name of the cookie.
+     * @param value The value of the cookie.
+     * @param maxAgeSeconds The max-age of the cookie in seconds. A value of 0 means delete the cookie.
+     * @param httpOnly If the cookie should be HttpOnly. Defaults to true.
+     * @param secure If the cookie should be Secure. Defaults to true.
+     * @return A formatted `Set-Cookie` string.
+     */
+    fun createCookieString(
+        name: String,
+        value: String,
+        maxAgeSeconds: Long,
+        httpOnly: Boolean = true,
+        secure: Boolean = true,
+    ): String {
+        val parts = mutableListOf("$name=$value", "Path=/", "Max-Age=$maxAgeSeconds", "SameSite=Lax")
+        if (httpOnly) parts.add("HttpOnly")
+        if (secure) parts.add("Secure")
+        return parts.joinToString("; ")
+    }
+}
+
+/**
+ * A set of public gRPC methods that do not require authentication.
+ */
+private val PUBLIC_METHODS =
+    setOf(
+        "snowballr.SnowballR/GetAuthenticationStatus",
+        "snowballr.SnowballR/Register",
+        "snowballr.SnowballR/Login",
+        "snowballr.SnowballR/RequestPasswordReset",
+        "snowballr.SnowballR/ResetPassword",
+    )
+
+/**
+ * Parses a "Cookie" header string into a map of key-value pairs.
+ * @param cookieHeader The string value of the "Cookie" header.
+ * @return A map of cookie names to cookie values.
+ */
+private fun parseCookies(cookieHeader: String?): Map<String, String> {
+    if (cookieHeader.isNullOrBlank()) return emptyMap()
+    return cookieHeader
+        .split(";")
+        .map { it.trim() }
+        .filter { it.contains('=') }
+        .associate {
+            val parts = it.split('=', limit = 2)
+            parts[0] to parts[1]
+        }
+}
+
+/**
+ * A [ServerInterceptor] for handling authentication via JWT.
+ *
+ * This interceptor performs the following steps:
+ * 1. Checks if the called method is public (e.g., login, register) or a health check.
+ *    If so, it bypasses authentication.
+ * 2. For protected methods, it extracts the access and refresh tokens from the gRPC metadata.
+ * 3. It validates the access token. If valid, it extracts the user ID, attaches it to the gRPC
+ *    [Context], and allows the call to proceed.
+ * 4. If the access token is invalid (e.g., expired), it attempts to use the refresh token to
+ *    generate a new access token.
+ * 5. If a new access token is successfully generated, it's sent back to the client in the
+ *    response trailers, and the call proceeds with the user ID attached to the context.
+ * 6. If authentication fails at any point (e.g., missing tokens, invalid refresh token),
+ *    the call is rejected with a [Status.UNAUTHENTICATED] status.
```

Comment:
Otherwise, the `[...]` parts aren't correctly translated to a reference in the comment.
```suggestion
 * 1. Checks if the called method is public (e.g., login, register) or a health check.
 * If so, it bypasses authentication.
 * 2. For protected methods, it extracts the access and refresh tokens from the gRPC metadata.
 * 3. It validates the access token. If valid, it extracts the user ID, attaches it to the gRPC
 * [Context], and allows the call to proceed.
 * 4. If the access token is invalid (e.g., expired), it attempts to use the refresh token to
 * generate a new access token.
 * 5. If a new access token is successfully generated, it's sent back to the client in the
 * response trailers, and the call proceeds with the user ID attached to the context.
 * 6. If authentication fails at any point (e.g., missing tokens, invalid refresh token),
 * the call is rejected with a [Status.UNAUTHENTICATED] status.
```

Reply:
Oh, you're right. I never noticed that, great catch!